tensorflow/lite/micro/kernels/xtensa/conv_hifi.cc

```diff
@@ -54,23 +54,142 @@ TfLiteStatus ConvPrepareHifi(TfLiteContext* context, TfLiteNode* node) {
   const int stride_height = params->stride_height;
   const int pad_height = data->reference_op_data.padding.height;
 
-  TF_LITE_ENSURE_EQ(context, input->type, kTfLiteInt8);
-
   int required_scratch = 0;
   // Dilation is currently not supported on HiFi 4 NN Library
   if ((params->dilation_width_factor == 1) &&
       (params->dilation_height_factor == 1)) {
-    required_scratch = xa_nn_conv2d_std_getsize(
-        input_height, input_depth, filter_height, filter_width, stride_height,
-        pad_height, output_height, output_channels, PREC_ASYM8S);
-    TF_LITE_ENSURE(context, required_scratch > 0);
+    if (input->type == kTfLiteInt8) {
+      required_scratch = xa_nn_conv2d_std_getsize(
+          input_height, input_depth, filter_height, filter_width, stride_height,
+          pad_height, output_height, output_channels, PREC_ASYM8S);
+      TF_LITE_ENSURE(context, required_scratch > 0);
+    }
+    if (input->type == kTfLiteInt16) {
+      required_scratch = xa_nn_conv2d_std_getsize(
+          input_height, input_depth, filter_height, filter_width, stride_height,
+          pad_height, output_height, output_channels, PREC_SYM16S);
+      TF_LITE_ENSURE(context, required_scratch > 0);
+    }
   }
   TF_LITE_ENSURE_OK(
       context, context->RequestScratchBufferInArena(
                    context, required_scratch, &data->scratch_tensor_index));
   return kTfLiteOk;
 }
 
+TfLiteStatus ConvEvalHifi16(TfLiteContext* context, TfLiteNode* node,
+                            const TfLiteConvParams& params,
+                            const XtensaConvOpData& data,
+                            const TfLiteEvalTensor* input,
+                            const TfLiteEvalTensor* filter,
+                            const TfLiteEvalTensor* bias,
+                            TfLiteEvalTensor* output) {
+  const RuntimeShape& input_shape = tflite::micro::GetTensorShape(input);
+  const RuntimeShape& filter_shape = tflite::micro::GetTensorShape(filter);
+  /* Dilation is currently not supported on HiFi 4 NN Library */
+  if ((params.dilation_width_factor == 1) &&
+      (params.dilation_height_factor == 1) &&
+      input_shape.Dims(1) >= filter_shape.Dims(1) &&
+      input_shape.Dims(2) >= filter_shape.Dims(2)) {
```

Comment:
this if-else block is quite confusing to read.

The else block has a return statement and then there is a call to the reference kernel after the else block. Refactoring to something like the following might help:

```cc
if (input_shape.Dims(1) < filter_shape.Dims(1)  || input_shape.Dims(2) >= filter_shape.Dims(2)) {
// first set of calls to nnlib
return
}

// Dilation not supported in nnlib
if ((params.dilation_width_factor == 1) &&
      (params.dilation_height_factor == 1) &&
      input_shape.Dims(1) >= filter_shape.Dims(1) &&
      input_shape.Dims(2) >= filter_shape.Dims(2)) {
 // calls to nnlib
  return;
}

// Use the reference kernels
}
```

I would also like know if the nnlib call when filter is large than the input is necessary, and whether that supports dilation or not.

Reply:
Created an cleanup bug http://b/210163787 to refactor it soon.